# The console through the ROM
#
# The MS 0515 has no DL11: keys come from the ROM's input entry and
# characters go to its output entry, the console registers are
# pseudo-registers at 300-306, the monitor interrupt (bit 8 of the system
# register) stands in for the output-ready interrupt, keys stay 8-bit for
# KOI-8, SO/SI switch the ROM's font at once, and the MS 7007 keyboard
# rows are released after each key and EMT (OMCONS.MAC).

diff --git a/KMON.MAC b/KMON.MAC
index 2f94950..74c2e1f 100644
--- a/KMON.MAC
+++ b/KMON.MAC
@@ -2539,6 +2539,7 @@ QREST1:	MOV	(PC)+,R1	;POINT TO QSIZE WORD		;BC17	;BC30
 	MOV	@(PC)+,-(SP)	;SET BIT ON FOR OUTPUT		;BC17	;BC30
 ..TTKS:: .WORD	TTKS-$RMON
 	BIS	#IENABL,@(SP)+	;ENABLE KEYBOARD INTERRUPTS
+	OMKBRL			;Omega: release the keyboard rows
 .ENDC ;EQ MTT$Y								;CG8-
 
 ; NOTE:	IF THIS IS THE MTTY SJ MONITOR, THE RTS BELOW WILL RETURN TO THE
diff --git a/RMONSJ.MAC b/RMONSJ.MAC
index 9cd2323..033c798 100644
--- a/RMONSJ.MAC
+++ b/RMONSJ.MAC
@@ -97,10 +97,10 @@ SYSVER:	.BYTE	RT$REL		;srr.vv rr part
 SYSUPD:	.BYTE	RT$VER		;srr.vv vv part
 CONFIG::.WORD	CLK50		;SYSTEM CONFIGURATION FLAGS
 SCROLL::.WORD	0		;LINK TO GT40 SCROLLER CODE
-TTKS::	.WORD	TKS		;CONSOLE IO ADDRESSES
-TTKB::	.WORD	TKB
-TTPS::	.WORD	TPS
-TTPB::	.WORD	TPB
+TTKS::	.WORD	CN$KS		;CONSOLE IO ADDRESSES (Omega: pseudo-registers)
+TTKB::	.WORD	CN$KB
+TTPS::	.WORD	CN$PS
+TTPB::	.WORD	CN$PB
 MAXBLK:	.WORD	-1		;MAXIMUM FILE SIZE TO ALLOW ON .ENTER
 E16LST:	.WORD	LST16-$RMON	;POINTER TO EMT 16 LIST FOR BATCH
 $TIME::	BSS	2		;DOUBLE PRECISION TIMER
@@ -690,12 +690,7 @@ POWRUP:									;005
 .ENDC ;NE PWF$L
 
 .IF NE	PWF$L!RDF$L
-CRASHP:	TSTB	@TTPS		;* TT BUSY
-	BPL	CRASHP		;* BR IF SO
-	MOVB	(R0)+,@TTPB	;* PRINT A CHARACTER
-	BNE	CRASHP		;* BR IF MORE TO PRINT
-2$:	HALT
-	BR	2$		;* AND NO WAY OUT			;HJ24-
+CRASHP:	OMCRSH			;Omega: through the ROM
 .ENDC ;NE PWF$L!RDF$L							;DV51-
 
 .DSABL	LSB
@@ -1017,6 +1012,7 @@ E16XT:	CLR	R2		;NO USER ARGS. TO POP
 ; THIS IS THE GENERAL END OF THE LINE FOR EMT'S
 
 EMTDON::JSR	R1,POPREG	;RESTORE A FEW REGISTERS
+	OMKBRL			;Omega: release the keyboard rows
 	RTI			;RETURN FROM EMT
 .SBTTL	TTYOUT
 
@@ -1192,19 +1188,17 @@ PTIBF4:: .WORD	TTIBUF+4	;*** BOOT ***
 	SPL	0		;* BACK TO LEVEL 0			;DV15
   .IFF  ;EQ TIME$R
 TTIINT::JSR	R5,$INTEN	;* DO COMMON INTERRUPT ENTRY
-	 .WORD	^C<PR4>&PR7	; (LEVEL 4)
+	 .WORD	^C<PR7>&PR7	; (Omega: LEVEL 7)
 	MOV	R3,-(SP)	;SAVE REGISTERS 3 - 0
 	MOV	R2,-(SP)
 	MOV	R1,-(SP)
 	MOV	R0,-(SP)
 	MOV	(PC)+,R5	;POINT TO RING COUNT
 PTIBF4:: .WORD	TTIBUF+4	;*** BOOT ***
-	MOV	@TTKB,R0	;CHARACTER INTO R0
+	OMTTIN			;Omega: the key from the ROM
   .ENDC ;EQ TIME$R							;LP2
  .ENDC ;NE HSR$B							;LP18
 
-	BIC	#^C<177>,R0	;STRIP OFF PARITY BIT			;LP12
-
  .IF NE	VT11$!VS60$							;DV57
 	JSR	PC,@SCLNK2	;MAYBE CALL SCROLLER
  .ENDC ;NE VT11$!VS60$
@@ -1308,6 +1302,7 @@ TTIEX3:
   .IF EQ	TIME$R
 	JMP	E16XT
   .IFF ;EQ TIME$R							;005
+	OMTTIX			;Omega: the next key, if any
 	MOV	(SP)+,R0	;RESTORE REGISTERS 0 - 3
 	MOV	(SP)+,R1
 	MOV	(SP)+,R2
@@ -1388,8 +1383,7 @@ TTCRLF:	JSR	PC,TTLFCR	;NEW LINE AFTER ^U
 
 CTRL.Q:	CLR	SYNC		;RESUME OUTPUT
 .IF EQ PDT$OP
-	BIC	#TPS$MK,@TTPS	;CLEAR INTERRUPT ENABLE
-	BIS	#IENABL,@TTPS	;START INTERRUPTING
+	OMTTON			;Omega: START INTERRUPTING
 .IFF
 	CLR	@TTPS		;CLEAR INTERRUPT ENABLE
 	MOV	#IENABL,@TTPS	;START INTERRUPTING
@@ -1458,7 +1452,7 @@ ZAP::									;005
 	RESET			;STOP ALL IO
 
 .IF EQ	PDT$OP
-	BIS	(SP)+,@TTPS	;RESET DLART CONSOLE TO PROGRAMMED BAUD RATE
+	OMRSET			;Omega: the output interrupt as it was
 .ENDC
 .IF EQ MTT$Y
 	CLR	SYNC		;Clear output stall flag
@@ -1611,7 +1605,7 @@ OPUT:	TSTB	TTOCLO		;IS ^O ON?
 	INC	-(R2)		;BUMP COUNT OF CHARACTERS IN BUFFER	;DG03
 	INC	2(R2)		;BUMP 'PUT' POINTER			;JH12-/DG03
 .IF EQ PDT$OP
-	BIS	#IENABL,@TTPS	;TURN ON THE OUTPUT INTERRUPT
+	OMTTON			;Omega: TURN ON THE OUTPUT INTERRUPT
 .IFF
 	MOV	#IENABL,@TTPS	;TURN ON THE OUTPUT INTERRUPT
 .ENDC
@@ -1641,7 +1635,7 @@ OPUT1:	TSTB	TTOCLO		;IS ^O ON?
 	INC	2(R2)		;BUMP 'PUT' POINTER			;JH12-/DG03
 	SPL	0		;return priority to what it was
 .IF EQ PDT$OP
-	BIS	#IENABL,@TTPS	;TURN ON THE OUTPUT INTERRUPT
+	OMTTON			;Omega: TURN ON THE OUTPUT INTERRUPT
 .IFF
 	MOV	#IENABL,@TTPS	;TURN ON THE OUTPUT INTERRUPT
 .ENDC
@@ -1658,7 +1652,7 @@ OPUT1:	TSTB	TTOCLO		;IS ^O ON?
 
 ; *** FOLLOWING LINE MODIFIED BY GT ON ***
 
-TTSCRL::BIC	#^C<177>,R4	;NO PARITY BITS
+TTSCRL::BIC	#^C<377>,R4	;Omega: eight bits, KOI-8
 	MOV	(PC)+,R2	;POINT TO COUNT OF BYTES IN OUTPUT BUFFER
 PTOBF2:: .WORD	TTOBUF+2	;*** BOOT ***
 	RTS	PC
@@ -1681,11 +1675,12 @@ CHCNT:	 .WORD	10		;# SPACES TO NEXT TAB STOP
 
 .IF NE	TIME$R
 TTOINT::JSR	R5,$INTEN	;* DO COMMON INTERRUPT ENTRY
-	 .WORD	^C<PR4>&PR7	; (LEVEL 0)
+	 .WORD	^C<PR7>&PR7	; (Omega: LEVEL 7)
 	MOV	(PC)+,R5	;POINT TO COUNT
 PTOBF4:: .WORD	TTOBUF+4	;*** BOOT ***
 	MOV	(PC)+,R4	;POINT TO TABCNT
 PTBCNT:: .WORD	TABCNT		;*** BOOT ***
+	OMTTOF			;Omega: the interrupt taken
 .ENDC ;NE TIME$R							;LP2
 TTO2:	TST	(PC)+		;ARE WE DOING OUTPUT?
 SYNC:	 .WORD	0
@@ -1707,7 +1702,7 @@ SYNC:	 .WORD	0
 	CMP	(R5)+,#-1	;DONE YET?
 	BEQ	11$		;*** WE DO THIS SO TT: CAN USE THIS CODE ALSO
 	MOVB	@(R5)+,@R4	;GET CHARACTER, SAVE IT IN FILLC
-3$:	BIC	#^C<177>,@R4	;STRIP PARITY				;JH9+
+3$:	BIC	#^C<377>,@R4	;Omega: eight bits, KOI-8
 	CMP	(R4)+,#RUBOUT	;Delete character?			;011
 	BEQ	4$		;Yes					;011
 	CMP	-2(R4),#40	;Test for printing character		;011
@@ -1717,7 +1712,7 @@ SYNC:	 .WORD	0
 	MOV	#10,@R4		;RESET TAB COUNT
 4$:	TST	-(R4)		;POINT TO CHARACTER TO PRINT IN FILLC
 5$:	INC	-(R5)		;BUMP OUTPUT RING POINTER
-6$:	MOV	@R4,@TTPB	;PRINT CHARACTER
+6$:	OMTTPB			;Omega: PRINT CHARACTER
 	BR	12$
 									;005
 7$:	CMP	-2(R4),#BS	;Backspace character?			;011
@@ -1738,7 +1733,7 @@ SYNC:	 .WORD	0
 10$:	DEC	R4		;BACK UP POINTER
 	CMPB	-(R4),@#TTFILL	;HAVE TO FILL?
 	BNE	5$		;NO. JUST PRINT THIS ONE
-	MOV	@R4,@TTPB	;TYPE THIS ONE
+	OMTTPB			;Omega: TYPE THIS ONE
 	CLR	@R4		;FILLERS ARE NULLS
 	CLR	-(R4)		;POINT R1 TO FILL COUNT AND CLEAR IT
 	MOVB	@#TTNFIL,@R4	;XFER FILL COUNT TO TABCNT
@@ -1747,12 +1742,14 @@ SYNC:	 .WORD	0
 
 11$:
 .IF EQ PDT$OP
-	BIC	#TPS$MK,@TTPS	;CLEAR INTERRUPTS
+	OMTTOF			;Omega: CLEAR INTERRUPTS
 .IFF
 	CLR	@TTPS		;CLEAR INTERRUPTS
 .ENDC
 	CLR	-(R5)
-12$:
+	BR	13$		;Omega
+12$:	OMTTRA			;Omega: ask for the next one
+13$:
 .IF EQ	TIME$R								;LP2
 	BR	RTICMN
 
diff --git a/USR.MAC b/USR.MAC
index bd90c52..75b0815 100644
--- a/USR.MAC
+++ b/USR.MAC
@@ -622,6 +622,7 @@ RSTIOW	== .			;RESET WAIT FOR I/O TO FINISH
 .TTKS	== .+2		;**LOAD** RELOCATE POINTER TO KEYBOARD STATUS ADDRESS
 	MOV	@#TTKS-$RMON,-(SP) ;SET BIT ON FOR OUTPUT
 	BIS	#IENABL,@(SP)+
+	OMKBRL			;Omega: release the keyboard rows
 .ENDC ;EQ MTT$Y
 	RTS	PC
 
